src/Swarm/TUI/Controller.hs

```diff
@@ -616,55 +620,58 @@ handleREPLEvent s = \case
+            Nothing -> uiState %= resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
             Just found
-              | T.null t -> s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
-              | otherwise ->
-                s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt found)
-                  & validateREPLForm
-      else continueWithoutRedraw s
-  Key V.KUp -> continue $ s & adjReplHistIndex Older
-  Key V.KDown -> continue $ s & adjReplHistIndex Newer
-  ControlKey 'r' -> continue $
+              | T.null t -> uiState %= resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
+              | otherwise -> do
+                uiState %= resetWithREPLForm (mkReplForm $ mkCmdPrompt found)
+                modify validateREPLForm
+      else continueWithoutRedraw
+  Key V.KUp -> modify $ adjReplHistIndex Older
+  Key V.KDown -> modify $ adjReplHistIndex Newer
+  ControlKey 'r' -> do
+    s <- get
     case s ^. uiState . uiReplForm . to formState of
       CmdPrompt uinput _ ->
         let newform = mkReplForm $ SearchPrompt uinput (s ^. uiState . uiReplHistory)
-         in s & uiState . uiReplForm .~ newform
+         in uiState . uiReplForm .= newform
       SearchPrompt ftext rh -> case lastEntry ftext rh of
-        Nothing -> s
+        Nothing -> pure ()
         Just found ->
           let newform = mkReplForm $ SearchPrompt ftext (removeEntry found rh)
-           in s & uiState . uiReplForm .~ newform
+           in uiState . uiReplForm .= newform
   CharKey '\t' -> do
-    let formSt = s ^. uiState . uiReplForm . to formState
-        newform = mkReplForm $ tabComplete s formSt
-    continue $
-      s & uiState . uiReplForm .~ newform
-        & validateREPLForm
-  EscapeKey ->
-    case s ^. uiState . uiReplForm . to formState of
-      CmdPrompt {} -> continueWithoutRedraw s
+    formSt <- use $ uiState . uiReplForm . to formState
+    newform <- gets $ mkReplForm . flip tabComplete formSt
+    uiState . uiReplForm .= newform
+    modify validateREPLForm
+  EscapeKey -> do
+    formSt <- use $ uiState . uiReplForm . to formState
+    case formSt of
+      CmdPrompt {} -> continueWithoutRedraw
       SearchPrompt _ _ ->
-        continue $ s & uiState %~ resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
+        uiState %= resetWithREPLForm (mkReplForm $ mkCmdPrompt "")
   ev -> do
-    f' <- handleFormEvent ev (s ^. uiState . uiReplForm)
-    continue $
-      case formState f' of
-        CmdPrompt {} -> validateREPLForm (s & uiState . uiReplForm .~ f')
-        SearchPrompt t _ ->
-          let newform = set promptUpdateL t (s ^. uiState)
-           in s & uiState . uiReplForm .~ newform
+    replForm <- use $ uiState . uiReplForm
+    f' <- nestEventM' replForm (handleFormEvent ev)
+    case formState f' of
+      CmdPrompt {} -> do
+        uiState . uiReplForm .= f'
+        modify validateREPLForm
+      SearchPrompt t _ -> do
+        -- TODO: why does promptUpdateL not update the uiState?
```

Comment:
Yes, `promptUpdateL` has a strange type, I am not sure why it is that way.

Reply:
Well, the only other place where its called is this:
```Haskell
uiState %= resetWithREPLForm (set promptUpdateL "" (s ^. uiState))
```
Honestly, I do not want to touch that `promptUpdateL` lens in this PR anymore.
It could cause more problems and I would not be sure what caused them. :fearful:

Reply:
I agree, we should not try to change it here.  Just something to think about for the future.